### DISCUSSION TITLE
Dynamically Changing Theme

### DISCUSSION THREAD

**Discussion by leadhkr** [tinymce/tinymce-react#207] [Q&A]
It's great that tinymce supports dark/light mode and it works beautifully. Is there a way to make this change dynamic without having to refresh the page? I have a global var that flips from light to dark theme as the user pushes a button. That choice doesn't reload the editor when changed (whether there's a state change or not).

Thanks

**Comment by ikokollari** [comment]
This is to spare anyone else from wasting hours trying to achieve this.

If you follow [Create a skin for TinyMCE](https://www.tiny.cloud/docs/tinymce/latest/creating-a-skin/) and browse the `less` theme
you will realize that there is no easy way to create a skin that would respond, say, to a `.dark` class change, or implement css variables.

The skin css is injected in the document head, while the content_css depends whether the editor is rendered inline or iframe.

One clunky solution for me was to replace the injected css when the theme changed like:

> [!WARNING]
> PSEUDOCODE

```javascript
useEffect(() => {
   document.head.querySelectorAll("[id^='someId']").remove();
   const link = document.createElement('link');
  link.href = 'url-to-skin.css';
  document.head.append(link);
}, [theme]);
```
With this method you soon realize that different skins have more than one stylesheet, so ....

I settled with re-initializing the editor anytime the theme changed. Like the following:
> [!CAUTION]
> This will destroy the `UndoManager` state
> And this was only tested [Using the TinyMCE React component as a uncontrolled component](https://www.tiny.cloud/docs/tinymce/latest/react-ref/#using-the-tinymce-react-component-as-a-uncontrolled-component)
```javascript
const Editor: React.FC = () => {

    const editorRef = useRef<TinyMCEEditor | null>(null);
    const { theme } = useTheme();

    const init = useMemo(() => {
        return {
            ...defaultSettings,
            skin: theme === 'dark' ? 'oxide-dark' : 'oxide',
            content_css: theme === 'dark' ? 'dark' : 'default',
        }
    }, [theme]);

    useEffect(() => {
        if (editorRef.current) {
            const editor = editorRef.current;
            const content = editor.getContent();

            const editorID = editor.id;
            tinymce.execCommand('mceRemoveEditor', false, { id: editorID });
            tinymce.execCommand('mceAddEditor', false, { id: editorID, options: init });

            editorRef.current = tinymce.get(editorID)!;
            editorRef.current.setContent(content);
        }
    }, [theme, init])

return <TinyMCEComponent
    onInit=(evt, editor) => editorRef.current = editor}
    init={init}
    ref={ref}
    tinymceScriptSrc='node_modules/tinymce/tinymce.min.js' // link to where you are hosting your copy of tinymce.
    licenseKey='gpl'
/>;
}
```

Where `theme: 'dark' | 'light'`

There is some flickering, but this is more time proof. And hopefully you only have one instance of the editor that needs reloading :)

**Comment by dollyzn** [comment]
I only added `key={theme}` to the Editor. When the theme changes from dark to light, etc., it automatically reloads the editor component.

**Reply by hsdev29032004** [reply]
it's not work

**Reply by dollyzn** [reply]
> it's not work

the `{theme}` must match your current system theme and be mutable. when the theme changes, the variable must change too.

![Gravando 2025-09-15 152650](https://github.com/user-attachments/assets/52229f16-0346-4403-9689-c7856322bf41)

**Reply by ikokollari** [reply]
Doesn't this destroy and re-initialize the editor, loosing data?

**Reply by dollyzn** [reply]
> Isso não destrói e reinicializa o editor, perdendo dados?

only if you put your content state inside the editor component, otherwise you just lose the undo/redo and save state
